### ISSUE TITLE
Express CLI

### ISSUE THREAD

**Issue by waleedtariq109** [expressjs/discussions#197]
First of all, a huge appreciation to the Express team for coming together and reviving Express.

**My suggestion is:**
Express should release its CLI, which scaffolds the project boilerplate. While Express itself is a framework, the current setup process requires starting from scratch, leading to a less-than-ideal experience. Let's discuss the need for a cleaner, more robust, and modern version of the Express CLI, going beyond the capabilities of express-generator.

**Comment by wesleytodd** [comment]
Hey! I agree. We have the generator package but it is quite old and doesn't follow the `create-*` pattern which the package managers support either. I think it would be awesome change the name of that, but that likely might mean a full rewrite. Maybe the next best steps here are to list the requirements?

**Comment by waleedtariq109** [comment]
@wesleytodd I completely agree, and I would also like to join the Express team for creating and maintaining the brand-new CLI package

**Comment by jonchurch** [comment]
I previously was excited about the idea of having the cli generator, but see the value less now. Id be interested to see your ideas about what it should do. 

Express is so lean that Im not certain what we would need to scaffold for folks. Previously the generator focused on scaffolding a server side rendered static html site using view engines. That's not as relevant today. 

Before you begin work @waleedtariq109 can you give an idea of what your vision is for a cli package?

**Comment by UlisesGascon** [comment]
> Express is so lean that Im not certain what we would need to scaffold for folks. Previously the generator focused on scaffolding a server side rendered static html site using view engines. That's not as relevant today.

AFAIK, with the current roadmap, I share this vision too. But here's a random idea from my side: what if we just offer a CLI that lists all the projects in the current example folder and allows you to choose one of them to start a project? This way, you actually have tons of options out of the box. Maybe I'm going too far; it's just an idea.

**Comment by wesleytodd** [comment]
Even if we get the most simple setup, I think the value of a generator for new users cannot be over stated. It is much different to say "create a file and copy this" vs `npm init express`.

**Comment by waleedtariq109** [comment]
@wesleytodd @jonchurch 
Just to provide a brief overview of my thoughts: Instead of installing Express using `npm i express`, it's recommended to use the Express CLI for scaffolding the project. The official way to do this is by executing the command `npx express init my-app`. This command will initialize an Express app. Additionally, it will prompt you with a few questions before initializing the app, such as whether you want to set up a Git repository and which package manager you prefer to use, among other things.

This has to be the Offical way of express

**Comment by ctcpip** [comment]
> follow the `create-*` pattern which the package managers support

what is this referring to?

**Comment by wesleytodd** [comment]
> what is this referring to?

package managers have a feature to download and execute packages with the naming pattern `create-`. For example:

```
$ npm init exress
```

Will find the package `create-express` or `@express/create` and will run the properly named bin script just like `npx` would.

**Comment by wesleytodd** [comment]
> such as whether you want to set up a Git repository and which package manager you prefer to use, among other things etc.

And then for this part, I think we should use the pattern I have been working out here: https://github.com/pkgjs/create-pkg/issues/7

**Comment by jonchurch** [comment]
@ctcpip see the `npm init` docs: https://docs.npmjs.com/cli/v8/commands/npm-init

**Comment by waleedtariq109** [comment]
> > such as whether you want to set up a Git repository and which package manager you prefer to use, among other things etc.
> 
> And then for this part, I think we should use the pattern I have been working out here: [pkgjs/create-pkg#7](https://github.com/pkgjs/create-pkg/issues/7)

Exactly, I also think the same way. We should use this pattern to provide a better DX, and I am also willing and passionate to contribute as a member. 😊

**Comment by waleedtariq109** [comment]
@wesleytodd I also want to participate in any discussion regarding this package. If there's another communication channel available, I'd be happy to join, as it's difficult to know if there are new messages here.

**Comment by wesleytodd** [comment]
Not too early at all! I think my personal focus is on the pressing work around v5, security, governance, and LTS strategy but that doesn't block anything. Just might mean I personally cannot help much yet. I think updating this issue to be that requirements and plan would be good. Or if this conversation is already too long you could do it as a new issue.

**Comment by jonchurch** [comment]
I think gathering requirements is the stage we are in right now. The idea of having express itself expose a bin so we can do `npx express <something>` is interesting to me, but is a big departure from how things are now.

For context, [express-generator](https://www.npmjs.com/package/express-generator) gets less than 9k downloads weekly currently. Which is not zero, but effectively rounds to it when you consider express gets 29 million downloads per week. Meaning, I don't think the generator is holding back any users, or presenting them with issues we need to address with priority. 

I'd love to see a proposal of what problems a CLI would solve (I think there are many it could tackle!), and the actual implementation of it being an `npm init` or `npx express` pattern is a detail which is tangential to those problems we'd want to solve.

**Comment by wesleytodd** [comment]
> The idea of having express itself expose a bin so we can do npx express <something> is interesting to me, but is a big departure from how things are now.

I personally am opposed to this at first glance. I will hold that weekly, but also I think the package we ship to prod servers **should not include more than the requirements for runtime**.

> For context, [express-generator](https://www.npmjs.com/package/express-generator) gets less than 9k downloads weekly currently.

Yeah this is more of an indication of it's failure to be updated than it's usefulness in general imo.

**Comment by jonchurch** [comment]
It is indeed trending downward, [peaked at around 800k/year](https://npm-stat.com/charts.html?package=express-generator&from=2014-01-28&to=2024-02-28) back in 2018 ish if these numbers are accurate:

<img width="979" alt="Screenshot 2024-02-28 at 1 31 41 PM" src="https://github.com/expressjs/discussions/assets/12915163/035d10e6-4090-47fc-8da6-6a326c4a58de">

**Comment by crandmck** [comment]
> For context, [express-generator](https://www.npmjs.com/package/express-generator) gets less than 9k downloads weekly currently. Which is not zero, but effectively rounds to it when you consider express gets 29 million downloads per week.

Obviously, Express itself is extremely popular--no one is going to argue with that, but aren't "downloads" of express/express here (i.e. from npm stats) counted every time anyone does an npm/npx update or install of any app with an Express dependency?  OTOH, express-generator is a CLI build-time tool, so those are actual installations of the CLI (not run-time deps), which makes the two statistics not really comparable.  I'm no npm expert, so perhaps I'm misunderstanding this?

Anyway, I don't think we should get too hung up on the numbers of downloads.  I just want to second what @wesleytodd said:

> Even if we get the most simple setup, I think the value of a generator for new users cannot be over stated. It is much different to say "create a file and copy this" vs npm init express.

I think having a easy tool for new users is very valuable, even if intermediate to advanced devs don't use it and it doesn't cover many use cases.  As long as it covers some basic use cases and patterns, it has value IMO.

**Comment by waleedtariq109** [comment]
@crandmck Yup, I totally agree with you. It is a valuable tool for new users. Of course, we don't have to cover every use case, but we should include all the basic operations like creating a route, using middleware with CLI, etc. Although it may sound like no one uses these commands except for `npm express init` but I think it's a great value addition.

**Comment by bjohansebas** [comment]
Hi team, I've created a repository to work on the design of the options that this new CLI could have (it won't have much logic for now). I'm taking inspiration from how Vite and Next.js work when creating a new application. I'll be finishing it in the next few days.

Repo: https://github.com/bjohansebas/create-express

**Comment by bjohansebas** [comment]
Hi

Maybe we should review this in future meetings, @kjugi has messaged me privately and is interested in helping, just like with the codemod repository.

Btw, i haven't followed up much with the repository i mentioned above because i've been involved in other projects, but if there's consensus, i can focus on it and help launch this new CLI for Express.

**Comment by bjohansebas** [comment]
I added the tc-meeting label to discuss it and make a decision about the package

**Comment by marcelino-borges** [comment]
Bringind another discussion I had opened to here:

I have created this CLI inspired in NestJS' CLI:
https://www.npmjs.com/package/expresst-cli

Github:
https://github.com/marcelino-borges/expresst-cli

I have just published a pre release version, but it can already be used in any project.

No need to install it as dependency, just call `npx expresst-cli`.

It has commands to initialize a full boilerplate with husky, git, commitlint, swagger (pick what you want) or add an adapter, use case, repository, route, controller, service etc. in an existing project.

It would be really nice to have some comments, suggestions, feedbacks on this.

**Comment by IgorArnaut** [comment]
> Hey! I agree. We have the generator package but it is quite old and doesn't follow the `create-*` pattern which the package managers support either. I think it would be awesome change the name of that, but that likely might mean a full rewrite. Maybe the next best steps here are to list the requirements?

You did not improve the generator. You did not put ES6 option. You did not put Typescript option. Put those features already.

**Comment by IgorArnaut** [comment]
@waleedtariq109 Your generator is 6 YEARS OLD.

**Comment by bjohansebas** [comment]
Okay, I finished the CLI I was talking about. It can still be improved, but that can be done if the package gets accepted.

You can try it with:

```sh
npm create @bjohansebas
```
the repo: https://github.com/bjohansebas/create-express/settings

**Comment by jonchurch** [comment]
After digging into https://github.com/bjohansebas/create-express I think that it takes on complexity in places where the payoff doesn't seem worth it. As in, there is a lot of combinatorial complexity w/ the exposed options, but the options are in places that are low value. It lets folks pick between 4 test runners and 4 linters and also has a ESM => CJS regex transform.

I think the real value is giving folks something with "express-y" patterns built in that they can follow the form of to continue to develop their app from the starter, and things that everyone will need like logging, error handling, security defaults.

I'd rather ship a few complete, opinionated starters than a config wizard. The examples in https://github.com/bjohansebas/create-express are good places to start.Other generators like create-vite etc often take that route, a small-ish number of stubbed examples, instead of being able to create over 1500 unique combinations like the https://github.com/bjohansebas/create-express repo currently allows.

**Comment by jonchurch** [comment]
Jean took a stab here https://github.com/sheplu/eg
